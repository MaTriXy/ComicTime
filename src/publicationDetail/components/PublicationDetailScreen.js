import React from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router";
import NavigationBar from "../../baseComponents/navigationBar/NavigationBar";
import { translateComponent } from "../../i18n/i18n";
import ProgressBar from "../../baseComponents/progressBar/ProgressBar";
import { Scrollbars } from "react-custom-scrollbars";
import { downloadIssue, fetchPublicationById } from "../publicationDetail";
import { None } from "monet";
import { List, ListItem, Paper } from "material-ui";
import LazyImage from "../../baseComponents/lazyImage/LazyImage";
import {
  gridTemplateForTwoColumns,
  imageBackgroundPositionForTwoColumns
} from "../../utils/windowUtils";
import ArrowDownward from "material-ui/svg-icons/navigation/arrow-downward";
import PublicationNotFound from "./PublicationNotFound";

class PublicationDetailScreen extends React.Component {
  constructor(props) {
    super(props);
    this.updateDimensions = this.updateDimensions.bind(this);
    this.sizeStatus = this.sizeStatus.bind(this);
  }

  componentWillMount() {
    const publicationId = this.props.match.params.id;
    this.props.onComponentMounted(publicationId);
    this.updateDimensions();
  }

  componentDidMount() {
    window.addEventListener("resize", this.updateDimensions);
  }

  componentWillUnmount() {
    window.removeEventListener("resize", this.updateDimensions);
  }

  updateDimensions() {
    this.setState(this.sizeStatus());
  }

  sizeStatus() {
    return { width: window.innerWidth, height: window.innerHeight };
  }

  componentDidUpdate(prevProps) {
    const oldPublicationId = prevProps.match.params.id;
    const publicationId = this.props.match.params.id;
    if (oldPublicationId !== publicationId) {
      this.props.onComponentMounted(publicationId);
    }
  }

  render() {
    if (this.props.loading) {
      return <ProgressBar />;
    } else {
      const style = this.style();
      const maybePublication = this.props.publication;
      const publicationTitle = maybePublication
        .map(publication => publication.title)
        .orSome("");
      const mainComponent = maybePublication
        .map(publication => {
          return this.publicationDetailComponent(style, publication);
        })
        .orSome(<PublicationNotFound />);
      return (
        <div style={{ height: "100%", width: "100%", position: "absolute" }}>
          <NavigationBar showBackButton={true} title={publicationTitle} />
          {mainComponent}
        </div>
      );
    }
  }

  publicationDetailComponent(style, publication) {
    const { t } = this.props;
    const publicationImage = publication.image;
    const publicationTitle = publication.title;
    return (
      <Scrollbars style={style.content}>
        <Paper style={style.paper}>
          <div style={style.infoGrid}>
            <div style={{ gridRowStart: 1, gridRowEnd: 7 }}>
              <LazyImage
                height={350}
                src={publicationImage}
                placeholder={publicationImage}
                alt={publicationTitle}
                style={style.image}
              />
            </div>
            <div>
              <p>
                <strong style={style.remarkableInfo}>
                  {t("publicationDetailHeaderTitle")}
                </strong>{" "}
                {publication.title}
              </p>
            </div>
            <div>
              <p>
                <strong style={style.remarkableInfo}>
                  {t("publicationDetailHeaderStatus")}
                </strong>{" "}
                {publication.status.orSome("")}
              </p>
            </div>
            <div>
              <p>
                <strong style={style.remarkableInfo}>
                  {t("publicationDetailHeaderSummary")}
                </strong>{" "}
                {publication.summary}
              </p>
            </div>
            <div>
              <p>
                <strong style={style.remarkableInfo}>
                  {t("publicationDetailHeaderPublisher")}
                </strong>{" "}
                {publication.publisher.orSome("")}
              </p>
            </div>
            <div>
              <p>
                <strong style={style.remarkableInfo}>
                  {t("publicationDetailHeaderReleaseDate")}
                </strong>{" "}
                {publication.releaseDate.orSome("")}
              </p>
            </div>
            <div>
              <p>
                <strong style={style.remarkableInfo}>
                  {t("publicationDetailHeaderAuthors")}
                </strong>{" "}
                {publication.authors.orSome("")}
              </p>
            </div>
          </div>
          <List>
            {publication.issues.map(issue => {
              return (
                <ListItem
                  key={issue.id}
                  primaryText={issue.title}
                  secondaryText={issue.releaseDate}
                  rightIconButton={
                    <ArrowDownward
                      onClick={() => {
                        this.props.onIssueClick(issue);
                      }}
                    />
                  }
                  onClick={() => {
                    this.props.onIssueClick(issue);
                  }}
                />
              );
            })}
          </List>
        </Paper>
      </Scrollbars>
    );
  }

  style() {
    return {
      content: {
        marginTop: 62
      },
      paper: {
        width: "90%",
        marginLeft: "auto",
        marginRight: "auto",
        marginBottom: 20,
        marginTop: 20,
        padding: 20
      },
      infoGrid: {
        display: "grid",
        gridTemplateColumns: gridTemplateForTwoColumns(this.state.width),
        marginBottom: 20
      },
      image: {
        backgroundPosition: imageBackgroundPositionForTwoColumns(
          this.state.width
        )
      },
      remarkableInfo: {
        textDecoration: "underline"
      }
    };
  }
}

PublicationDetailScreen.defaultProps = {
  loading: true,
  publication: None(),
  error: None()
};

export function mapStateToProps(state) {
  return {
    loading: state.publicationDetail.fetchingPublication,
    publication: state.publicationDetail.publicationSelected,
    error: state.publicationDetail.errorFetchingPublication
  };
}

export function mapPropsToDispatch(dispatch) {
  return {
    onComponentMounted: publicationId => {
      dispatch(fetchPublicationById(publicationId));
    },
    onIssueClick: issue => {
      dispatch(downloadIssue(issue));
    }
  };
}

export const PurePublicationDetailScreen = PublicationDetailScreen;
export default withRouter(
  connect(mapStateToProps, mapPropsToDispatch)(
    translateComponent(PurePublicationDetailScreen)
  )
);
