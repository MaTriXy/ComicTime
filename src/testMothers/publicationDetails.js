import { IssueSummary, Publication } from "../publicationDetail/model";
import { Issue, IssuePage } from "../issueDetail/model";
import { Some } from "monet";

export const anyPublicationId = "the-flash-2016";
export const anyIssueId = "Annual1";

export const anyPublication = new Publication(
  "the-flash-2016",
  "https://comicimages.com/the-flash-1",
  "The Flash",
  Some("Ongoing"),
  "The best comic series in the world",
  Some("DC"),
  Some("January 2016"),
  Some("Peter Gomez"),
  [
    new IssueSummary("1", "the-flash-2016", "The Flash #1", "January 2016"),
    new IssueSummary("2", "the-flash-2016", "The Flash #2", "February 2016")
  ]
);

export const anyIssueSummary = new IssueSummary(
  "1",
  "detective-comic",
  "Detective Comic #1",
  "January 2018"
);

export const anyIssue = new Issue(
  "1",
  "detective-comic",
  "Detective Comic #1",
  2,
  [
    new IssuePage(1, "https://comicpages.com/1"),
    new IssuePage(2, "https://comicpages.com/2")
  ]
);
