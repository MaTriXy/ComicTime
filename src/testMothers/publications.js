import { Page, PublicationSummary } from "../feed/model";

export const anyListOfPublications = [
  new PublicationSummary(
    "uuid1",
    "http://readcomicsonline.ru/comic/deadpool-2017",
    "deadpool-2017",
    "deadpool (2017)",
    ["deadpool (2017) #1"],
    "http://readcomicsonline.ru/uploads/manga/deadpool-2017/cover/cover_250x350.jpg",
    "20/1/2017",
    "Marvel"
  ),
  new PublicationSummary(
    "uuid2",
    "http://readcomicsonline.ru/comic/volcanosaurus-2017",
    "volcanosaurus-2017",
    "Volcanosaurus (2017)",
    ["Volcanosaurus (2017) #2"],
    "http://readcomicsonline.ru/uploads/manga/volcanosaurus-2017/cover/cover_250x350.jpg",
    "04/5/2018",
    "American Mythology"
  )
];

export const anyFirstPageOfPublications = new Page(anyListOfPublications, 1, 1);
export const anySecondPageOfPublications = new Page(
  anyListOfPublications,
  2,
  2
);
