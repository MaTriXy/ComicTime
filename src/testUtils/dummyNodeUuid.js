jest.mock("node-uuid", () => {
  return {
    v4() {
      return "uuid";
    }
  };
});
