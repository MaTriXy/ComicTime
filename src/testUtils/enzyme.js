import React from "react";
import Enzyme from "enzyme";
import { createSerializer } from "enzyme-to-json";
import toJson from "enzyme-to-json";
import Adapter from "enzyme-adapter-react-16";
import { MuiThemeProvider } from "material-ui";
import comicTimeTheme from "../theme";

expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));
Enzyme.configure({ adapter: new Adapter() });

export function mockHistory() {
  return {
    go: jest.fn(),
    goBack: jest.fn()
  };
}

export function translator(key) {
  return key;
}

export function mountWithMuyThemeProvider(component) {
  return Enzyme.mount(
    <MuiThemeProvider muiTheme={comicTimeTheme}>{component}</MuiThemeProvider>
  );
}

export function mountComponent(component) {
  return Enzyme.mount(component);
}

export function mountComponentAsJson(component) {
  return toJson(mountComponent(component));
}

export function shallowComponent(component) {
  return Enzyme.shallow(component);
}

export function shallowComponentAsJson(component) {
  return toJson(shallowComponent(component));
}

export function json(component) {
  return toJson(component);
}
