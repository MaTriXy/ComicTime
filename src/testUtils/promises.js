export async function flushPromises() {
  return new Promise(resolve => setImmediate(resolve));
}
