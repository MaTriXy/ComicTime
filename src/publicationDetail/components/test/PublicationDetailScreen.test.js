jest.mock("../../../i18n/i18n");
jest.mock("../PublicationNotFound", () => "PublicationNotFound");
jest.mock(
  "../../../baseComponents/navigationBar/NavigationBar",
  () => "NavigationBar"
);
jest.mock("material-ui/List", () => "List");
jest.mock("material-ui/List/ListItem", () => "ListItem");
jest.mock("material-ui/Paper", () => "Paper");

import React from "react";
import { shallowComponentAsJson, translator } from "../../../testUtils/enzyme";
import { PurePublicationDetailScreen } from "../PublicationDetailScreen";
import { Some, None } from "monet";
import { anyPublication } from "../../../testMothers/publicationDetails";
const match = {
  params: {
    id: anyPublication.id
  }
};

describe("PublicationDetailScreen", () => {
  it("shows the progress bar while the publication is fetched", () => {
    const component = shallowComponentAsJson(
      <PurePublicationDetailScreen
        loading={true}
        match={match}
        onComponentMounted={jest.fn()}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it("shows the publication not found component if there was an error loading the publication by id", () => {
    const component = shallowComponentAsJson(
      <PurePublicationDetailScreen
        loading={false}
        publication={None()}
        error={Some("")}
        match={match}
        onComponentMounted={jest.fn()}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it("shows the publication information if it was loaded properly", () => {
    const component = shallowComponentAsJson(
      <PurePublicationDetailScreen
        t={translator}
        loading={false}
        publication={Some(anyPublication)}
        error={None()}
        match={match}
        onComponentMounted={jest.fn()}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
