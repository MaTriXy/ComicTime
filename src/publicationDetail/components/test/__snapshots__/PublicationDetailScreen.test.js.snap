// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PublicationDetailScreen shows the progress bar while the publication is fetched 1`] = `
<ProgressBar
  hidden={false}
/>
`;

exports[`PublicationDetailScreen shows the publication information if it was loaded properly 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <NavigationBar
    showBackButton={true}
    title="The Flash"
  />
  <Scrollbars
    autoHeight={false}
    autoHeightMax={200}
    autoHeightMin={0}
    autoHide={false}
    autoHideDuration={200}
    autoHideTimeout={1000}
    hideTracksWhenNotNeeded={false}
    renderThumbHorizontal={[Function]}
    renderThumbVertical={[Function]}
    renderTrackHorizontal={[Function]}
    renderTrackVertical={[Function]}
    renderView={[Function]}
    style={
      Object {
        "marginTop": 62,
      }
    }
    tagName="div"
    thumbMinSize={30}
    universal={false}
  >
    <Paper
      style={
        Object {
          "marginBottom": 20,
          "marginLeft": "auto",
          "marginRight": "auto",
          "marginTop": 20,
          "padding": 20,
          "width": "90%",
        }
      }
    >
      <div
        style={
          Object {
            "display": "grid",
            "gridTemplateColumns": "35% 65%",
            "marginBottom": 20,
          }
        }
      >
        <div
          style={
            Object {
              "gridRowEnd": 7,
              "gridRowStart": 1,
            }
          }
        >
          <LazyImage
            alt="The Flash"
            height={350}
            placeholder="https://comicimages.com/the-flash-1"
            src="https://comicimages.com/the-flash-1"
            style={
              Object {
                "backgroundPosition": "0% 0%",
              }
            }
          />
        </div>
        <div>
          <p>
            <strong
              style={
                Object {
                  "textDecoration": "underline",
                }
              }
            >
              publicationDetailHeaderTitle
            </strong>
             
            The Flash
          </p>
        </div>
        <div>
          <p>
            <strong
              style={
                Object {
                  "textDecoration": "underline",
                }
              }
            >
              publicationDetailHeaderStatus
            </strong>
             
            Ongoing
          </p>
        </div>
        <div>
          <p>
            <strong
              style={
                Object {
                  "textDecoration": "underline",
                }
              }
            >
              publicationDetailHeaderSummary
            </strong>
             
            The best comic series in the world
          </p>
        </div>
        <div>
          <p>
            <strong
              style={
                Object {
                  "textDecoration": "underline",
                }
              }
            >
              publicationDetailHeaderPublisher
            </strong>
             
            DC
          </p>
        </div>
        <div>
          <p>
            <strong
              style={
                Object {
                  "textDecoration": "underline",
                }
              }
            >
              publicationDetailHeaderReleaseDate
            </strong>
             
            January 2016
          </p>
        </div>
        <div>
          <p>
            <strong
              style={
                Object {
                  "textDecoration": "underline",
                }
              }
            >
              publicationDetailHeaderAuthors
            </strong>
             
            Peter Gomez
          </p>
        </div>
      </div>
      <div
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        <a
          href="https://panels.app"
          rel="noopener noreferrer"
          target="_blank"
        >
          <img
            alt="https://panels.app"
            height="60"
            src="panelsLogo.svg"
            target="_blank"
          />
        </a>
      </div>
      <p
        style={
          Object {
            "textAlign": "center",
          }
        }
      >
        Download any issue from the issues lists and read it using
         
        <a
          href="https://panels.app"
          target="_blank"
        >
          Panels
        </a>
        , the best comic reader in the market.
         
      </p>
      <List>
        <ListItem
          key="1"
          onClick={[Function]}
          primaryText="The Flash #1"
          rightIconButton={
            <NavigationArrowDownward
              onClick={[Function]}
            />
          }
          secondaryText="January 2016"
        />
        <ListItem
          key="2"
          onClick={[Function]}
          primaryText="The Flash #2"
          rightIconButton={
            <NavigationArrowDownward
              onClick={[Function]}
            />
          }
          secondaryText="February 2016"
        />
      </List>
    </Paper>
  </Scrollbars>
</div>
`;

exports[`PublicationDetailScreen shows the publication not found component if there was an error loading the publication by id 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <NavigationBar
    showBackButton={true}
    title=""
  />
  <PublicationNotFound />
</div>
`;
